File: draft-order-complete-drops-voucher/node/detect-dropped-voucher.js
/**
 * Detect Saleor draft orders whose voucher discount was dropped or shrunk by
 * draftOrderComplete's price recalculation, and report them for finance review.
 *
 * Saleor recalculates an order's prices through its pricing manager at several
 * trigger points, including the transition draftOrderComplete performs. The
 * discount is re-derived from the order's stored voucher and voucherCode
 * reference every time that recalculation runs, and that path has historically
 * failed to consistently re-derive it, dropping the OrderDiscount linkage or
 * recomputing it against the wrong base.
 *
 * There is no safe auto-fix: re-adding a discount after completion can desync
 * the order total from an already-captured Transaction or Payment. This is
 * detect and report, with an optional orderDiscountAdd call gated by DRY_RUN
 * and meant to run only after a human has reviewed the finding.
 *
 * Guide: https://www.allanninal.dev/saleor/draft-order-complete-drops-voucher/
 */
import { pathToFileURL } from "node:url";

const API_URL = process.env.SALEOR_API_URL || "https://demo.saleor.io/graphql/";
const TOKEN = process.env.SALEOR_AUTH_TOKEN || "token_dummy";
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";

export function diffVoucherDiscount(draftSnapshot, completedSnapshot, tolerance = 0.01) {
  const expectedDiscount = draftSnapshot.undiscountedTotalGross - draftSnapshot.totalGross;
  const actualDiscount = completedSnapshot.undiscountedTotalGross - completedSnapshot.totalGross;
  const delta = expectedDiscount - actualDiscount;

  const voucherWasRemoved = Boolean(draftSnapshot.voucherCode) && !completedSnapshot.voucherCode;
  const discountShrank = delta > tolerance;

  const isDropped =
    Boolean(draftSnapshot.voucherCode) &&
    expectedDiscount > tolerance &&
    (voucherWasRemoved || discountShrank);

  return { isDropped, expectedDiscount, actualDiscount, delta };
}

async function gql(query, variables = {}) {
  const res = await fetch(API_URL, {
    method: "POST",
    headers: { Authorization: `Bearer ${TOKEN}`, "Content-Type": "application/json" },
    body: JSON.stringify({ query, variables }),
  });
  if (!res.ok) throw new Error(`Saleor ${res.status}`);
  const body = await res.json();
  if (body.errors) throw new Error(JSON.stringify(body.errors));
  return body.data;
}

const ORDER_SNAPSHOT_QUERY = `
query($id: ID!) {
  order(id: $id) {
    id
    status
    voucherCode
    voucher { code }
    discounts { type valueType value amount { amount } }
    total { gross { amount } }
    undiscountedTotal { gross { amount } }
  }
}`;

const COMPLETE_MUTATION = `
mutation($id: ID!) {
  draftOrderComplete(id: $id) {
    order { id }
    errors { field message }
  }
}`;

const DISCOUNT_ADD_MUTATION = `
mutation($orderId: ID!, $value: PositiveDecimal!, $reason: String!) {
  orderDiscountAdd(orderId: $orderId, input: { valueType: FIXED, value: $value, reason: $reason }) {
    order { id }
    errors { field message }
  }
}`;

async function fetchOrderSnapshot(orderId) {
  const { order } = await gql(ORDER_SNAPSHOT_QUERY, { id: orderId });
  return {
    voucherCode: order.voucherCode,
    totalGross: order.total.gross.amount,
    undiscountedTotalGross: order.undiscountedTotal.gross.amount,
  };
}

async function completeDraftOrder(draftId) {
  const result = (await gql(COMPLETE_MUTATION, { id: draftId })).draftOrderComplete;
  if (result.errors.length) throw new Error(JSON.stringify(result.errors));
  return result.order.id;
}

async function recoverDiscount(orderId, expectedDiscount) {
  const result = (
    await gql(DISCOUNT_ADD_MUTATION, {
      orderId,
      value: Math.round(expectedDiscount * 100) / 100,
      reason: "Recovered voucher discount from draft order snapshot",
    })
  ).orderDiscountAdd;
  if (result.errors.length) throw new Error(JSON.stringify(result.errors));
  return result.order.id;
}

export async function run(draftOrderIds) {
  const mode = DRY_RUN ? "dry run" : "live";
  console.log(`Checking ${draftOrderIds.length} draft order(s) for dropped vouchers (${mode})`);

  let flagged = 0;
  for (const draftId of draftOrderIds) {
    const draftSnapshot = await fetchOrderSnapshot(draftId);
    const completedId = await completeDraftOrder(draftId);
    const completedSnapshot = await fetchOrderSnapshot(completedId);

    const result = diffVoucherDiscount(draftSnapshot, completedSnapshot);
    if (!result.isDropped) continue;

    flagged++;
    console.warn(
      `Voucher dropped on order=${completedId} expected=${result.expectedDiscount.toFixed(2)} actual=${result.actualDiscount.toFixed(2)} delta=${result.delta.toFixed(2)}`
    );

    if (!DRY_RUN) await recoverDiscount(completedId, result.expectedDiscount);
  }

  console.log(
    `Done. ${flagged} order(s) with a dropped voucher ${DRY_RUN ? "to review" : "had a discount re-added"}.`
  );
  return flagged;
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run([]).catch((err) => { console.error(err); process.exit(1); });
}

File: draft-order-complete-drops-voucher/python/detect_dropped_voucher.py
"""Detect Saleor draft orders whose voucher discount was dropped or shrunk by
draftOrderComplete's price recalculation, and report them for finance review.

Saleor recalculates an order's prices through its pricing manager at several
trigger points, including the transition draftOrderComplete performs. The
discount is re-derived from the order's stored voucher and voucherCode
reference every time that recalculation runs, and that path has historically
failed to consistently re-derive it, dropping the OrderDiscount linkage or
recomputing it against the wrong base.

There is no safe auto-fix: re-adding a discount after completion can desync
the order total from an already-captured Transaction or Payment. This is
detect and report, with an optional orderDiscountAdd call gated by DRY_RUN
and meant to run only after a human has reviewed the finding.

Guide: https://www.allanninal.dev/saleor/draft-order-complete-drops-voucher/
"""
import os
import logging
import requests

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("detect_dropped_voucher")

API_URL = os.environ["SALEOR_API_URL"]
TOKEN = os.environ["SALEOR_AUTH_TOKEN"]
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"

ORDER_SNAPSHOT_QUERY = """
query($id: ID!) {
  order(id: $id) {
    id
    status
    voucherCode
    voucher { code }
    discounts { type valueType value amount { amount } }
    total { gross { amount } }
    undiscountedTotal { gross { amount } }
  }
}"""

COMPLETE_MUTATION = """
mutation($id: ID!) {
  draftOrderComplete(id: $id) {
    order { id }
    errors { field message }
  }
}"""

DISCOUNT_ADD_MUTATION = """
mutation($orderId: ID!, $value: PositiveDecimal!, $reason: String!) {
  orderDiscountAdd(orderId: $orderId, input: { valueType: FIXED, value: $value, reason: $reason }) {
    order { id }
    errors { field message }
  }
}"""


def gql(query, variables=None):
    r = requests.post(
        API_URL,
        json={"query": query, "variables": variables or {}},
        headers={"Authorization": f"Bearer {TOKEN}", "Content-Type": "application/json"},
        timeout=30,
    )
    r.raise_for_status()
    body = r.json()
    if body.get("errors"):
        raise RuntimeError(body["errors"])
    return body["data"]


def fetch_order_snapshot(order_id):
    order = gql(ORDER_SNAPSHOT_QUERY, {"id": order_id})["order"]
    return {
        "voucherCode": order["voucherCode"],
        "totalGross": order["total"]["gross"]["amount"],
        "undiscountedTotalGross": order["undiscountedTotal"]["gross"]["amount"],
    }


def diff_voucher_discount(draft_snapshot, completed_snapshot, tolerance=0.01):
    """
    Pure decision logic, no I/O.
    draft_snapshot / completed_snapshot: {"voucherCode": str | None,
                                           "totalGross": float,
                                           "undiscountedTotalGross": float}
    Returns {"isDropped": bool, "expectedDiscount": float,
             "actualDiscount": float, "delta": float}
    """
    expected_discount = draft_snapshot["undiscountedTotalGross"] - draft_snapshot["totalGross"]
    actual_discount = completed_snapshot["undiscountedTotalGross"] - completed_snapshot["totalGross"]
    delta = expected_discount - actual_discount

    voucher_was_removed = bool(draft_snapshot["voucherCode"]) and not completed_snapshot["voucherCode"]
    discount_shrank = delta > tolerance

    is_dropped = (
        bool(draft_snapshot["voucherCode"])
        and expected_discount > tolerance
        and (voucher_was_removed or discount_shrank)
    )

    return {
        "isDropped": is_dropped,
        "expectedDiscount": expected_discount,
        "actualDiscount": actual_discount,
        "delta": delta,
    }


def complete_draft_order(draft_id):
    result = gql(COMPLETE_MUTATION, {"id": draft_id})["draftOrderComplete"]
    if result["errors"]:
        raise RuntimeError(result["errors"])
    return result["order"]["id"]


def recover_discount(order_id, expected_discount):
    result = gql(
        DISCOUNT_ADD_MUTATION,
        {
            "orderId": order_id,
            "value": round(expected_discount, 2),
            "reason": "Recovered voucher discount from draft order snapshot",
        },
    )["orderDiscountAdd"]
    if result["errors"]:
        raise RuntimeError(result["errors"])
    return result["order"]["id"]


def run(draft_order_ids):
    mode = "dry run" if DRY_RUN else "live"
    log.info("Checking %d draft order(s) for dropped vouchers (%s)", len(draft_order_ids), mode)

    flagged = 0
    for draft_id in draft_order_ids:
        draft_snapshot = fetch_order_snapshot(draft_id)
        completed_id = complete_draft_order(draft_id)
        completed_snapshot = fetch_order_snapshot(completed_id)

        result = diff_voucher_discount(draft_snapshot, completed_snapshot)
        if not result["isDropped"]:
            continue

        flagged += 1
        log.warning(
            "Voucher dropped on order=%s expected=%.2f actual=%.2f delta=%.2f",
            completed_id, result["expectedDiscount"], result["actualDiscount"], result["delta"],
        )

        if not DRY_RUN:
            recover_discount(completed_id, result["expectedDiscount"])

    log.info(
        "Done. %d order(s) with a dropped voucher %s.",
        flagged, "to review" if DRY_RUN else "had a discount re-added",
    )
    return flagged


if __name__ == "__main__":
    run([])
